statistics/cmsketch.go

```diff
@@ -44,29 +58,215 @@ func NewCMSketch(d, w int32) *CMSketch {
 	return &CMSketch{depth: d, width: w, table: tbl}
 }
 
+// NewCMSketchWithTopN returns a new CM sketch with TopN elements.
+// Total is the size of the whole dataset
+// Only when the number of topn elements exceeded len(data) / topnThreshold will store topn elements
+func NewCMSketchWithTopN(d, w int32, data [][]byte, n uint32, total uint64) *CMSketch {
+	tbl := make([][]uint32, d)
+	for i := range tbl {
+		tbl[i] = make([]uint32, w)
+	}
+	c := &CMSketch{depth: d, width: w, table: tbl}
+	c.BuildTopN(data, n, topNThreshold, total)
+	return c
+}
+
+func groupElements(data [][]byte) map[uint64][]cmsCount {
+	counter := make(map[uint64][]cmsCount)
+
+	for k := range data {
+		h1, h2 := murmur3.Sum128(data[k])
+		vals, ok := counter[h1]
+		if !ok {
+			vals = make([]cmsCount, 0)
+		}
+		exists := false
+		for i := range vals {
+			if vals[i].h2 == h2 && bytes.Equal(vals[i].data, data[k]) {
+				exists = true
+				vals[i].count++
+			}
+		}
+		if !exists {
+			vals = append(vals, cmsCount{h1, h2, data[k], 1})
+		}
+		counter[h1] = vals
+	}
+
+	return counter
+}
+
+// BuildTopN builds table of top N elements.
+// elements in data should not be modified after this call.
+func (c *CMSketch) BuildTopN(data [][]byte, numTop, topNThreshold uint32, total uint64) {
+	counter := groupElements(data)
+	sorted := make([]uint64, 0)
+
+	for _, vals := range counter {
+		for i := range vals {
+			sorted = append(sorted, vals[i].count)
+		}
+	}
+	sort.Slice(sorted, func(i, j int) bool {
+		return sorted[i] > sorted[j]
+	})
+
+	sampleNDV := uint32(len(sorted))
+	if numTop > sampleNDV {
+		numTop = sampleNDV
+	}
+
+	NthValue := sorted[numTop-1]
+	newNthValue := sorted[numTop-1]
+	sumTopN := uint64(0)
+
+	// Add a few elements, the number of which is close to the n-th most element.
+	for i := uint32(0); i < sampleNDV && i < numTop*2; i++ {
+		// Here, 2/3 is get by running tests, tested 1, 1/2, 2/3, and 2/3 is relative better than 1 and 1/2
+		if i >= numTop && sorted[i]*3 < NthValue*2 && newNthValue != sorted[i] {
+			break
+		}
+		// sumTopN might be smaller than sum of final sum of elements in topNIndex.
+		// These two values are only used for build topNIndex, and they are not used in counting defaultValue.
+		newNthValue = sorted[i]
+		sumTopN += sorted[i]
+	}
+
+	estimateNDV, ratio, onlyOnceItems := calculateEstimateNDV(sorted, total)
+
+	sampleSize := uint64(len(data))
+	enableTopN := sampleSize/uint64(topNThreshold) <= sumTopN
+	topN := make([]cmsCount, 0)
+	sumTopN = 0
+
+	for _, vals := range counter {
+		for i := range vals {
+			if enableTopN && vals[i].count >= newNthValue {
+				topN = append(topN, cmsCount{data: vals[i].data, count: vals[i].count * ratio})
+				sumTopN += vals[i].count * ratio
+			} else {
+				c.insertBytesN(vals[i].data, vals[i].count*ratio)
+			}
+		}
+	}
+	numTop = uint32(len(topN))
+	if enableTopN {
+		c.buildTopNMap(topN)
+	}
+
+	countWithoutTopN := total - (sampleSize-uint64(onlyOnceItems))*ratio
```

Comment:
According to the variable's name, the formula should be `total - sumTopN*ratio`?

Reply:
Well, I found only use Top N will increase the error slightly,. And it is renamed now.